src/freenas/etc/periodic/security/800.loginfail-freenas

```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python
+#
+# This script is here because 800.loginfail does not check for log record year,
+# and auth.log can miss rotation and therefore include records from last year.
+# (see https://bugs.freenas.org/issues/18615)
+
+import bz2
+from datetime import datetime, timedelta
+import glob
+import gzip
+import os
+import re
+import sys
+
+
+def catmsgs():
+    log_directory = "/var/log"
+
+    for log_file in sorted(
+        filter(
+            lambda path: re.match(".*\.[0-9]+\.[^.]+$", path),
+            glob.glob("/var/log/auth.log.*.*"),
+        ),
+        key=lambda path: int(path.split(".")[-2]),
+        reverse=True,
+    ):
+        if log_file.endswith(".bz2"):
+            try:
+                with bz2.BZ2File(log_file, "rb") as f:
+                    yield from f
+            except IOError:
+                pass
+
+        if log_file.endswith(".gz"):
+            try:
+                with gzip.GzipFile(log_file, "rb") as f:
+                    yield from f
+            except IOError:
+                pass
+
+    try:
+        with open(os.path.join(log_directory, "auth.log"), "rb") as f:
+            yield from f
+    except IOError:
+        pass
+
+
+def get_login_failures(now, messages):
+    """
+    >>> get_login_failures(datetime(year=2017, month=8, day=31), [
+    ...    b"Aug 30 invalid login",  # 2017
+    ...    b"Aug 31 invalid login",  # 2017
+    ... ])
+    [b'Aug 30 invalid login']
+
+    >>> get_login_failures(datetime(year=2017, month=8, day=31), [
+    ...    b"Aug 30 invalid login",  # 2016
+    ...    b"Oct 18 invalid login",  # 2016
+    ...    b"Aug 31 invalid login",  # 2017
+    ... ])
+    []
+
+    >>> get_login_failures(datetime(year=2017, month=8, day=31), [
+    ...    b"Aug 30 invalid login",  # 2015
+    ...    b"Oct 18 invalid login",  # 2015
+    ...    b"Aug 31 invalid login",  # 2016
+    ...    b"Aug 30 bad login",      # 2017
+    ... ])
+    [b'Aug 30 bad login']
+    """
+
+    yesterday = (now - timedelta(days=1)).strftime("%b %e ").encode("ascii")
+    today = now.strftime("%b %e ").encode("ascii")
+
+    login_failures = []
+    for message in messages:
+        if message.startswith(yesterday):
+            if re.search(rb"\b(fail(ures?|ed)?|invalid|bad|illegal|auth.*error)\b", message):
+                login_failures.append(message)
+
+        if not message.startswith(yesterday) and not message.startswith(today):
+            login_failures = []
+
+    return login_failures
+
+if __name__ == "__main__":
+    # import doctest
+    # doctest.testmod()
+    # sys.exit(0)
+
+    # import pprint
+    # pprint.pprint(list(catmsgs()))
+    # sys.exit(0)
+
+    login_failures = get_login_failures(datetime.now(), catmsgs())
+    os.write(1, b"%s login failures:\n%s" % (
+        os.environ.get("host", "").encode("utf-8"),
+        b"".join(login_failures),
+    ))
```

Comment:
I am confused. Why are you writing to stdout using os.write?

Also, in the original script, the invalid logins are written to `stderr`, and  the number of entries follow it, in `stdout`.

Did I miss something?

Reply:
> Why are you writing to stdout using os.write?

Because I need to write `bytes` while `sys.stdout.write` accepts `str`

> Also, in the original script, the invalid logins are written to `stderr`

Oops, I've missed that. Fixed.

> and the number of entries follow it, in `stdout`

I don't see number of entries printed to `stdout`. Where?

```
if check_yesno_period security_status_loginfail_enable
then
	echo ""
	echo "${host} login failures:"
	n=$(catmsgs | egrep -ia "^$yesterday.*: .*\b(fail(ures?|ed)?|invalid|bad|illegal|auth.*error)\b" |
	    tee /dev/stderr | wc -l)
	[ $n -gt 0 ] && rc=1 || rc=0
fi

exit $rc
```